File: frontend/js/views/barista_fire.js
// Barista FIRE — portfolio covers the gap between expenses and a
// low-stress part-time income (canonical: Starbucks barista with
// healthcare). Smaller FI number than full FIRE.

import { api } from '../api.js';
import { esc } from '../util.js';
import { t } from '../i18n.js';

export async function renderBaristaFire(mount, _state) {
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.barista_fire.title">// BARISTA FIRE</span></h1>
        <p class="muted small" data-i18n-html="view.barista_fire.intro">
            <strong>Barista FI</strong> = portfolio is large enough to cover the gap
            between your living expenses and a low-stress part-time job's after-tax income.
            Canonical example: Starbucks-barista wages but with healthcare coverage. You can
            quit the high-stress career while the portfolio + part-time income carry you to
            traditional retirement. <code>barista_fi = (expenses − part_time_income) / SWR</code>.
        </p>
        <div class="chart-panel">
            <div style="display:grid;grid-template-columns:repeat(auto-fit,minmax(200px,1fr));gap:8px;margin-bottom:12px">
                <label><span class="muted small" data-i18n="view.barista_fire.field.nw">Current net worth $</span>
                    <input type="number" id="bf-nw" step="1000" min="0" value="200000" style="width:100%"></label>
                <label><span class="muted small" data-i18n="view.barista_fire.field.age">Current age</span>
                    <input type="number" id="bf-age" step="1" min="1" max="110" value="35" style="width:100%"></label>
                <label><span class="muted small" data-i18n="view.barista_fire.field.expenses">Annual expenses $</span>
                    <input type="number" id="bf-expenses" step="1000" min="0" value="40000" style="width:100%"></label>
                <label><span class="muted small" data-i18n="view.barista_fire.field.part_time">Annual part-time income $</span>
                    <input type="number" id="bf-pt" step="1000" min="0" value="25000" style="width:100%"></label>
                <label><span class="muted small" data-i18n="view.barista_fire.field.swr">SWR %</span>
                    <input type="number" id="bf-swr" step="0.25" min="0.25" max="20" value="4" style="width:100%"></label>
                <label><span class="muted small" data-i18n="view.barista_fire.field.return">Real return %/yr</span>
                    <input type="number" id="bf-return" step="0.25" min="-10" max="20" value="5" style="width:100%"></label>
                <label><span class="muted small" data-i18n="view.barista_fire.field.contrib">Monthly contribution $</span>
                    <input type="number" id="bf-contrib" step="50" min="0" value="500" style="width:100%"></label>
            </div>
            <button class="btn btn-sm primary" id="bf-run" data-shortcut="r" data-i18n="view.barista_fire.btn.run">⚡ Compute Barista FI</button>
            <div id="bf-result"></div>
        </div>
    `;
    mount.querySelector('#bf-run').addEventListener('click', () => runCompute(mount));
    await runCompute(mount);
}

async function runCompute(mount) {
    const result = mount.querySelector('#bf-result');
    const input = {
        current_nw_usd: parseFloat(mount.querySelector('#bf-nw').value) || 0,
        current_age: parseInt(mount.querySelector('#bf-age').value, 10) || 0,
        annual_expenses_usd: parseFloat(mount.querySelector('#bf-expenses').value) || 0,
        annual_part_time_income_usd: parseFloat(mount.querySelector('#bf-pt').value) || 0,
        safe_withdrawal_rate_pct: parseFloat(mount.querySelector('#bf-swr').value) || 4,
        expected_real_return_pct: parseFloat(mount.querySelector('#bf-return').value) || 5,
        current_monthly_contribution_usd: parseFloat(mount.querySelector('#bf-contrib').value) || 0,
    };
    result.innerHTML = `<p class="muted">${esc(t('view.barista_fire.status.computing'))}</p>`;
    try {
        const r = await api('/barista-fire/compute', { method: 'POST', body: JSON.stringify(input) });
        const stCls = r.status === 'barista_fi' ? 'pos' : r.status === 'no_gap' ? 'pos' : 'neg';
        const deltaCls = r.current_vs_barista_delta_usd >= 0 ? 'pos' : 'neg';
        result.innerHTML = `
            <div style="display:grid;grid-template-columns:repeat(auto-fit,minmax(200px,1fr));gap:12px;margin-top:1rem">
                <div><div class="muted small">${esc(t('view.barista_fire.field.full_fi'))}</div>
                    <strong>$${(r.full_fi_number_usd / 1000).toFixed(0)}K</strong></div>
                <div><div class="muted small">${esc(t('view.barista_fire.field.barista_fi'))}</div>
                    <strong style="font-size:1.4em">$${(r.barista_fi_number_usd / 1000).toFixed(0)}K</strong></div>
                <div><div class="muted small">${esc(t('view.barista_fire.field.savings'))}</div>
                    <strong class="pos">$${(r.barista_savings_usd / 1000).toFixed(0)}K</strong></div>
                <div><div class="muted small">${esc(t('view.barista_fire.field.gap'))}</div>
                    <strong>$${r.gap_annual_usd.toFixed(0)}/yr</strong></div>
                <div><div class="muted small">${esc(t('view.barista_fire.field.delta'))}</div>
                    <strong class="${deltaCls}">${r.current_vs_barista_delta_usd >= 0 ? '+' : '−'}$${(Math.abs(r.current_vs_barista_delta_usd) / 1000).toFixed(1)}K</strong></div>
                <div><div class="muted small">${esc(t('view.barista_fire.field.years_to'))}</div>
                    <strong>${r.years_until_barista_fi == null ? '∞' : r.years_until_barista_fi.toFixed(1) + 'y'}</strong></div>
                <div><div class="muted small">${esc(t('view.barista_fire.field.status'))}</div>
                    <strong class="${stCls}" style="text-transform:uppercase">${esc(t('view.barista_fire.status.' + r.status) || r.status)}</strong></div>
            </div>
        `;
    } catch (e) {
        result.innerHTML = `<p class="neg">${esc(String(e))}</p>`;
    }
}

File: crates/traderview-db/src/barista_fire.rs
//! Barista FIRE calculator.
//!
//! "Barista FI" is the point at which your portfolio is large enough
//! to cover the **gap** between your living expenses and a low-stress
//! part-time job's after-tax income — the canonical example being a
//! Starbucks barista job that pays modest wages but covers healthcare.
//! You can quit the high-stress career and let the portfolio + part-time
//! income carry you to traditional retirement.
//!
//! Math:
//!   gap_annual_usd = annual_expenses_usd − annual_part_time_income_usd
//!   barista_fi_now = gap_annual_usd / SWR_pct   (= a smaller FI number)
//!   coast_today    = barista_fi_now / (1+r)^years_until_retirement
//!     (similar discounting to coast-fire — at retirement age you stop
//!     working part-time too, so portfolio must cover FULL expenses then;
//!     barista_fi is a near-term-only number)
//!
//! Reports:
//!   - full_fi_number_usd   — what the user would need without part-time
//!   - barista_fi_number_usd — what they need with part-time income
//!   - barista_savings_usd   — full_fi − barista_fi
//!   - current_vs_barista_delta_usd
//!   - is_barista_fi flag
//!   - years_until_barista_fi at current contribution rate
//!
//! Pure compute.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct BaristaFireInput {
    pub current_nw_usd: f64,
    pub current_age: u32,
    pub annual_expenses_usd: f64,
    pub annual_part_time_income_usd: f64,
    #[serde(default = "default_swr")]
    pub safe_withdrawal_rate_pct: f64,
    #[serde(default = "default_return")]
    pub expected_real_return_pct: f64,
    #[serde(default)]
    pub current_monthly_contribution_usd: f64,
}

fn default_swr() -> f64 { 4.0 }
fn default_return() -> f64 { 5.0 }

#[derive(Debug, Clone, Serialize)]
pub struct BaristaFireReport {
    pub full_fi_number_usd: f64,
    pub barista_fi_number_usd: f64,
    pub barista_savings_usd: f64,
    pub gap_annual_usd: f64,
    pub current_vs_barista_delta_usd: f64,
    pub is_barista_fi: bool,
    pub years_until_barista_fi: Option<f64>,
    pub status: &'static str,
}

// ─── Pure compute ─────────────────────────────────────────────────────────

pub fn fi_number(annual_expenses: f64, swr_pct: f64) -> f64 {
    if swr_pct <= 0.0 { return 0.0; }
    annual_expenses / (swr_pct / 100.0)
}

/// Months to grow `current` + monthly contributions to `target` at
/// `annual_return_pct` real return. Returns None if never reaches in 100 years.
pub fn years_until_target(
    current: f64,
    monthly_contribution: f64,
    annual_return_pct: f64,
    target: f64,
) -> Option<f64> {
    if current >= target { return Some(0.0); }
    if monthly_contribution <= 0.0 && annual_return_pct <= 0.0 { return None; }
    let r = annual_return_pct / 100.0;
    let monthly_r = r / 12.0;
    let mut years = 0.0_f64;
    let step = 1.0 / 12.0;
    while years < 100.0 {
        years += step;
        let months = years * 12.0;
        let lump = current * (1.0 + r).powf(years);
        let annuity = if monthly_r.abs() < 1e-12 {
            monthly_contribution * months
        } else {
            monthly_contribution * ((1.0 + monthly_r).powf(months) - 1.0) / monthly_r
        };
        if lump + annuity >= target { return Some(years); }
    }
    None
}

pub fn compute(input: &BaristaFireInput) -> BaristaFireReport {
    let full_fi = fi_number(input.annual_expenses_usd, input.safe_withdrawal_rate_pct);
    let gap = (input.annual_expenses_usd - input.annual_part_time_income_usd).max(0.0);
    let barista_fi = fi_number(gap, input.safe_withdrawal_rate_pct);
    let savings = (full_fi - barista_fi).max(0.0);
    let delta = input.current_nw_usd - barista_fi;
    let is_barista = input.current_nw_usd >= barista_fi - 0.005;
    let years_to_barista = years_until_target(
        input.current_nw_usd,
        input.current_monthly_contribution_usd,
        input.expected_real_return_pct,
        barista_fi,
    );
    let _ = input.current_age;
    let status: &'static str = if barista_fi <= 0.0 { "no_gap" }
        else if is_barista { "barista_fi" }
        else { "not_yet" };
    BaristaFireReport {
        full_fi_number_usd: full_fi,
        barista_fi_number_usd: barista_fi,
        barista_savings_usd: savings,
        gap_annual_usd: gap,
        current_vs_barista_delta_usd: delta,
        is_barista_fi: is_barista,
        years_until_barista_fi: years_to_barista,
        status,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn input() -> BaristaFireInput {
        BaristaFireInput {
            current_nw_usd: 200_000.0,
            current_age: 35,
            annual_expenses_usd: 40_000.0,
            annual_part_time_income_usd: 25_000.0,
            safe_withdrawal_rate_pct: 4.0,
            expected_real_return_pct: 5.0,
            current_monthly_contribution_usd: 500.0,
        }
    }

    #[test]
    fn fi_number_basic() {
        assert_eq!(fi_number(40_000.0, 4.0), 1_000_000.0);
    }

    #[test]
    fn years_until_target_already_there() {
        assert_eq!(years_until_target(500_000.0, 0.0, 5.0, 100_000.0), Some(0.0));
    }

    #[test]
    fn years_until_target_finite() {
        let y = years_until_target(100_000.0, 1_000.0, 5.0, 500_000.0).unwrap();
        assert!(y > 5.0 && y < 25.0);
    }

    #[test]
    fn years_until_target_none_when_impossible() {
        assert!(years_until_target(100.0, 0.0, 0.0, 1_000_000.0).is_none());
    }

    #[test]
    fn compute_full_fi_basic() {
        let r = compute(&input());
        assert_eq!(r.full_fi_number_usd, 1_000_000.0);
    }

    #[test]
    fn compute_barista_fi_smaller_than_full() {
        let r = compute(&input());
        // gap = 40k − 25k = 15k. barista = 15k / 4% = 375k.
        assert_eq!(r.barista_fi_number_usd, 375_000.0);
        assert!(r.barista_fi_number_usd < r.full_fi_number_usd);
        assert_eq!(r.barista_savings_usd, 625_000.0);
    }

    #[test]
    fn compute_gap_zero_when_part_time_covers_expenses() {
        let mut i = input();
        i.annual_part_time_income_usd = 45_000.0;
        let r = compute(&i);
        assert_eq!(r.gap_annual_usd, 0.0);
        assert_eq!(r.barista_fi_number_usd, 0.0);
        assert_eq!(r.status, "no_gap");
    }

    #[test]
    fn compute_not_yet_at_200k() {
        let r = compute(&input());
        assert!(!r.is_barista_fi);
        assert!(r.current_vs_barista_delta_usd < 0.0);
        assert_eq!(r.status, "not_yet");
    }

    #[test]
    fn compute_already_barista_fi_at_400k() {
        let mut i = input();
        i.current_nw_usd = 400_000.0;
        let r = compute(&i);
        assert!(r.is_barista_fi);
        assert_eq!(r.status, "barista_fi");
        assert_eq!(r.years_until_barista_fi, Some(0.0));
    }

    #[test]
    fn compute_years_to_barista_positive_at_200k() {
        let r = compute(&input());
        // Need to grow 200k → 375k with $500/mo at 5% real. Should be < 20 years.
        assert!(r.years_until_barista_fi.is_some());
        let y = r.years_until_barista_fi.unwrap();
        assert!(y > 5.0 && y < 30.0);
    }
}
